teps/0071-custom-task-sdk.md

```diff
@@ -0,0 +1,220 @@
+---
+status: proposed
+title: Custom Task SDK
+creation-date: '2021-06-15'
+last-updated: '2021-06-15'
+authors:
+- '@ScrapCodes'
+---
+
+# TEP-0071: Custom Task SDK
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Notes/Caveats (optional)](#notescaveats-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+    - [User Experience (optional)](#user-experience-optional)
+    - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A Custom Task SDK will make the work of custom task author easier and will offer
+standard way to implement custom task controllers. While adhering to current
+separation of concern between a custom task controller's role and tektoncd owned
+controller's role, it should make custom tasks easier to manage and reason by
+tektoncd.
+
+## Motivation
+
+1. Currently, a custom task author has limited ways to know what is a standard way
+    of building his custom task. His best bet is to explore one of the existing 
+   implementation in `experimental/` repo.
+2. However, an experimental custom task may not be up to date with the latest `tektoncd`
+   code base version. There is no one goto place, where a Task author can find upto
+   date example.
+3. Presently, `tektoncd` does not manage the life cycle of a custom task at all.
+   A SDK can open up newer possibilities in this regard.
+4. SDK available as a project template, can perform some of common tasks,
+   generate boiler-plate code and provide helper functions/scripts.
+
+### Goals
+
+1. Provide a Github template with a simple controller and CRD implementation,
+   which a user can fork and develop his own custom task controller.
+2. Provide helper functions/packages for common tasks and documentation
+   with best practices.
+
+### Non-Goals
+
+TBD.
```

Comment:
ideas:
- providing SDKs in languages other than Go
- providing a generic CRD controller lib

Question: would this project involve us creating an maintaining any libraries for this, or would the intention at this point be just boiler plate that folks copy and paste? (either makes sense to me - im thinking it would be simplest to start with boilerplate and think about adding libs later, since libs will require their own maintenance?)

Reply:
Yes, I agree. Starting with the boiler plate and then moving towards libs. Disadvantage with boiler plate is, every time we update it, the end users may not get the update. Or the update can be at times non trivial change.